### ISSUE TITLE
[BUG] High CPU usage: "deviceinfochange" event is being dispatched indefinitely

### ISSUE THREAD

**Issue by cmaiquel** [twilio/twilio-voice.js#126]
- [x] I have verified that the issue occurs with the latest twilio.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-voice.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-voice.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/TwilioDevEd/voice-javascript-sdk-quickstart-node) application works in my environment.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Code to reproduce the issue:**

[mediadevices.js](https://github.com/twilio/twilio-voice.js/blob/master/lib/twilio/shims/mediadevices.js)
When a media device has an empty label, the method "deviceInfosHaveChanged" always returns **true**
A macOS "videoinput" device with an empty label, causing this behavior:

```
    {
        "deviceId": "",
        "kind": "videoinput",
        "label": "",
        "groupId": "zxzczxzccxcxxzczczxczxc"
    }
```

```js
// Always returns true if there's a media device with an empty label, even when there's no difference between newDevices and oldDevices
function deviceInfosHaveChanged(newDevices, oldDevices) {
    var oldLabels = oldDevices.reduce(function (map, device) { return map.set(device.deviceId, device.label || null); }, new Map());
    return newDevices.some(function (newDevice) {
        var oldLabel = oldLabels.get(newDevice.deviceId);
        return typeof oldLabel !== 'undefined' && oldLabel !== newDevice.label;
    });
}

// As a result 'deviceinfochange' is triggered each time this method is invoked
// and this method is being invoked in a "poll" interval of 500ms
function sampleDevices(mediaDevices) {
    nativeMediaDevices.enumerateDevices().then(function (newDevices) {
        var knownDevices = mediaDevices._knownDevices;
        var oldDevices = knownDevices.slice();
        // Replace known devices in-place
        [].splice.apply(knownDevices, [0, knownDevices.length]
            .concat(newDevices.sort(sortDevicesById)));
        if (!mediaDevices._deviceChangeIsNative
            && devicesHaveChanged(knownDevices, oldDevices)) {
            mediaDevices.dispatchEvent(new Event('devicechange'));
        }
        if (!mediaDevices._deviceInfoChangeIsNative
            && deviceInfosHaveChanged(known devices, oldDevices)) {
            mediaDevices.dispatchEvent(new Event('deviceinfochange'));
        }
    });
}
```

**Expected behavior:**

Do not dispatch "deviceinfochange" in an indefinite loop for empty labeled devices.

**Software versions:**

 - [x] Browser(s): Chrome Version 107.0.5304.110 (Official Build) (x86_64)
 - [x] Operating System: macOS 12.6.1
 - [x] twilio.js: 2.1.2
 - [ ] Third-party libraries (e.g., Angular, React, etc.):

**Comment by charliesantos** [comment]
Hi @cmaiquel , thanks for submitting. Usually devices will show with empty labels if you haven't asked for media permissions, or the user denied media access. We recommend to call getUserMedia and get the necessary permissions before enumerating devices or placing a call.

With that said, are you able to reproduce this issue after getting the necessary media permissions?

**Comment by cmaiquel** [comment]
@charliesantos I'm using the Voice SDK for voice calls. Ideally, I wouldn't need the video device permissions.

**Comment by charliesantos** [comment]
I submitted an internal ticket to address this @cmaiquel .

**Comment by cmaiquel** [comment]
Thanks @charliesantos.
There should also be a better way to detect device changes instead of calling enumerate devices every 500ms.
I think that "devicechange" event from MediaDevices interface is widely supported nowadays. https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/devicechange_event

**Comment by charliesantos** [comment]
@cmaiquel I agree! This should get addressed in our modernization effort for the SDK.

**Comment by kamalbennani** [comment]
I kinda agree with @cmaiquel.
We are also facing the same issue kind of, because of this shim device detection, which sometimes causes double-stream audio.

Lately, we moved the device management logic into a separate logic instead of using Twilio internal implementation, and we react to the device change based on the modern browser APIs, but the issue, is when we try to let the Twilio SDK know about a newly plugged device, the SDK will just crash and fail clean-up resources (cleaning up the master audio, etc ...), as the SDK is not yet aware of this new plugged Device because of this 500ms interval.

I think based on your Browser Support List, this most probably needs to be dropped off in favour of the native browser APIs.

**Comment by charliesantos** [comment]
I completely agree @kamalbennani @cmaiquel !

**Comment by charliesantos** [comment]
@cmaiquel @kamalbennani we started the work on this issue. Just letting you know that we are probably not going to drop the polling logic since some of our [supported browsers](https://www.twilio.com/docs/voice/sdks/javascript#supported-browsers) still does not [support](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/devicechange_event#browser_compatibility) the native API. However, we will figure out how to feature detect this one so we don't have to poll on browsers that do support the native API.

**Comment by charliesantos** [comment]
Should be fixed in https://github.com/twilio/twilio-voice.js/pull/155
However, we still need to keep the polling part to keep the internal deviceinfochange, since the native devicechange event is not emitted when device labels are updated. We are thinking of creating a parameter to allow disabling of polling to save more CPU, for those that really needs it.

**Comment by charliesantos** [comment]
Should be fixed in 2.4.0

**Comment by cmaiquel** [comment]
Thanks! I will try the fix.
